src/wayland_keybinds.cpp

```diff
@@ -1,138 +1,313 @@
+SeatInfo::SeatInfo(uint32_t name, wl_seat *seat)
+        : _name(name), _seat(seat), _keyboard(nullptr),
+          _xkb_state(nullptr), _pressed_syms()
+{
+    wl_seat_add_listener(seat, &seat_listener, this);
+}
 
-struct wl_registry_listener registry_listener {
-   .global = registry_handle_global,
-   .global_remove = registry_handle_global_remove
-};
+SeatInfo::~SeatInfo()
+{
+    fini_keyboard();
 
-struct wl_keyboard_listener keyboard_listener {
-   .keymap = wl_keyboard_keymap,
-   .enter = wl_keyboard_enter,
-   .leave = wl_keyboard_leave,
-   .key = wl_keyboard_key,
-   .modifiers = wl_keyboard_modifiers,
-   .repeat_info = wl_keyboard_repeat_info
-};
+    if (wl_seat_get_version(_seat) >= WL_SEAT_RELEASE_SINCE_VERSION)
+        wl_seat_release(_seat);
+    else
+        wl_seat_destroy(_seat);
+}
+
+bool SeatInfo::is_global(uint32_t name) const
+{
+    return _name == name;
+}
+
+bool SeatInfo::are_pressed(const std::vector<KeySym> &syms) const
+{
+    return syms == _pressed_syms;
+}
+
+void SeatInfo::init_keyboard()
+{
+    if (_keyboard)
+        return;
+
+    _keyboard = wl_seat_get_keyboard(_seat);
+    if (_keyboard)
+        wl_keyboard_add_listener(_keyboard, &keyboard_listener, this);
+}
+
+void SeatInfo::fini_keyboard()
+{
+    xkb_state_unref(_xkb_state);
+    _xkb_state = nullptr;
+
+    if (!_keyboard) {
+        return;
+    }
+
+    if (wl_keyboard_get_version(_keyboard) >= WL_KEYBOARD_RELEASE_SINCE_VERSION)
+        wl_keyboard_release(_keyboard);
+    else
+        wl_keyboard_destroy(_keyboard);
+
+    _keyboard = nullptr;
+}
+
+void SeatInfo::keymap(uint32_t format, int fd, size_t size)
+{
+    clear();
+
+    xkb_state_unref(_xkb_state);
+    _xkb_state = nullptr;
+
+    if (format != WL_KEYBOARD_KEYMAP_FORMAT_XKB_V1) {
+        close(fd);
+        return;
+    }
+
+    void *map_shm = mmap(nullptr, size, PROT_READ, MAP_PRIVATE, fd, 0);
+    close(fd);
+    if (map_shm == MAP_FAILED)
+        return;
+
+    xkb_keymap *keymap = xkb_keymap_new_from_string(
+            xkb_context,
+            static_cast<char *>(map_shm),
+            XKB_KEYMAP_FORMAT_TEXT_V1,
+            XKB_KEYMAP_COMPILE_NO_FLAGS);
+    munmap(map_shm, size);
 
-static void seat_handle_capabilities(void *data, wl_seat *seat, uint32_t caps)
+    if (keymap) {
+        _xkb_state = xkb_state_new(keymap);
+        xkb_keymap_unref(keymap);
+    }
+}
+
+void SeatInfo::key(uint32_t key, uint32_t state)
 {
-   if(caps & WL_SEAT_CAPABILITY_KEYBOARD)
-   {
-      if(!keyboard)
-      {
-         keyboard = wl_seat_get_keyboard(seat);
-         wl_keyboard_add_listener(keyboard, &keyboard_listener, NULL);
-      }
-   }
+    if (!_xkb_state)
+        return;
+
+    xkb_keycode_t code = key + 8;
+    xkb_keysym_t sym = xkb_state_key_get_one_sym(_xkb_state, code);
+    if (sym == XKB_KEY_NoSymbol)
+        return;
+
+    if (state == WL_POINTER_BUTTON_STATE_PRESSED) {
```

Comment:
this should probably be WL_KEYBOARD_KEY_STATE_PRESSED

Reply:
Ah yes, whoops